From b9652445b121adf4738b8a0003f6701504778b1c Mon Sep 17 00:00:00 2001
From: Ethan Wei <ethanwei@ami.com>
Date: Thu, 7 Sep 2023 23:38:34 -0700
Subject: [PATCH] Re-Design the RA part in DHCPEnabled. 1. When DHCPv6 enabled,
 set acceptRA to true. 2. When DHCPv6 disabled, remain current value of
 acceptRA

	modified:   src/ethernet_interface.cpp
---
 src/ethernet_interface.cpp | 16 +++++++++++-----
 1 file changed, 11 insertions(+), 5 deletions(-)

diff --git a/src/ethernet_interface.cpp b/src/ethernet_interface.cpp
index 325e914..a096f6f 100644
--- a/src/ethernet_interface.cpp
+++ b/src/ethernet_interface.cpp
@@ -488,13 +488,19 @@ EthernetInterface::DHCPConf EthernetInterface::dhcpEnabled(DHCPConf value)
     auto old6 = EthernetInterfaceIntf::dhcp6();
     auto new6 = EthernetInterfaceIntf::dhcp6(value == DHCPConf::v6 ||
                                              value == DHCPConf::both);
-    auto oldra = EthernetInterfaceIntf::ipv6AcceptRA();
-    auto newra = EthernetInterfaceIntf::ipv6AcceptRA(
-        value == DHCPConf::v6stateless || value == DHCPConf::v4v6stateless ||
-        value == DHCPConf::v6 || value == DHCPConf::both);
 
-    if (old4 != new4 || old6 != new6 || oldra != newra)
+    if (old4 != new4 || old6 != new6)
     {
+        if (EthernetInterfaceIntf::dhcp6() && old6 != new6) {
+            EthernetInterfaceIntf::ipv6AcceptRA(true);
+            EthernetInterfaceIntf::ipv6EnableStaticRtr(false);
+            if (!EthernetInterfaceIntf::ipv6StaticRtrAddr().empty()) {
+                if (auto it = staticNeighbors.find(stdplus::fromStr<stdplus::In6Addr>(EthernetInterfaceIntf::ipv6StaticRtrAddr())); it != staticNeighbors.end()) {
+                    staticNeighbors.erase(stdplus::fromStr<stdplus::In6Addr>(EthernetInterfaceIntf::ipv6StaticRtrAddr()));
+                    EthernetInterfaceIntf::ipv6StaticRtrAddr(std::string{});
+                }
+            }
+        }
         writeConfigurationFile();
         manager.get().reloadConfigs();
     }
-- 
2.25.1

